allennlp_models/generation/dataset_readers/cnn_dm_tt.py

```diff
@@ -0,0 +1,136 @@
+import random
+from typing import Dict, Optional, List, Any, Iterable
+
+from datasets import load_dataset, Dataset
+from overrides import overrides
+import torch.distributed as dist
+
+from allennlp.common import cached_transformers, util
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.fields import Field, TransformerTextField
+from allennlp.data.instance import Instance
+
+
+@DatasetReader.register("cnn_dm_tt")
+class CNNDailyMailDatasetReaderTransformerToolkit(DatasetReader):
+    """
+    A CNN/Daily dataset reader for models that use the transformer toolkit.
+    """
+
+    def __init__(
+        self,
+        transformer_model_name: str,
+        source_length_limit: int = 512,
+        target_length_limit: int = 56,
+        source_prefix: Optional[str] = None,
+        shuffle: bool = False,
+        num_preprocessing_workers: Optional[int] = None,
+        **kwargs,
+    ) -> None:
+        super().__init__(
+            manual_distributed_sharding=True, manual_multiprocess_sharding=True, **kwargs
+        )
+        self._tokenizer = cached_transformers.get_tokenizer(transformer_model_name)
+        self._source_length_limit = source_length_limit
+        self._target_length_limit = target_length_limit
+        self._source_prefix = source_prefix
+        self._shuffle = shuffle
+        self._num_preprocessing_workers = num_preprocessing_workers
+
+    @overrides
+    def _read(self, split: str):
+        dataset: Dataset = load_dataset("cnn_dailymail", version="3.0.0", split=split)  # type: ignore
+        progress_prefix = ""
+
+        if util.is_distributed():
+            dataset = dataset.shard(dist.get_world_size(), dist.get_rank())
+
+        if self._worker_info is not None and self._worker_info.num_workers > 1:
+            dataset = dataset.shard(self._worker_info.num_workers, self._worker_info.id)
+            progress_prefix = f"[worker {self._worker_info.id}] "
```

Comment:
This will shard twice?

Reply:
Yea, once for distributed, once for multi-process loading within the same distributed worker.